### PR TITLE
Added metrics collection with plan validation and DB locking

### PR DESCRIPTION
**Pull request by andreasanta**
Refactor of web-server code inside handlers and implementation of metrics collection with DB transactions

### CODE REVIEW

File: iot-meter/webhandlers/login_handler.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+		
+		// TEMPORARY CODE TO GENERATE JWT
+		if (email == "admin" && password == "admin") {
+			return &models.Account{
+				ID  : "MASTER",
+				Email : "admin@goteleport.com",
+				PlanType : "S",
+				TotalUsers : 0,
+			}, nil
+		}
```
Comment (russjones) [inline iot-meter/webhandlers/login_handler.go:34]: Looks like `go fmt` was not run on this file, same with other files. Seeing odd indentation.

File: iot-meter/webhandlers/main_server.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    "net/http"
+)
+
+const JwtIdKey = "iot_account_id"
+
+type MainServer struct {
+    DB     *gorm.DB
+    Router *gin.Engine
+
+    _authMiddleware *jwt.GinJWTMiddleware
```
Comment (russjones) [inline iot-meter/webhandlers/main_server.go:23]: Not idiomatic go to have variables with underscores in them, you can just call this `authMiddleware`.

File: iot-meter/webhandlers/metrics_handler.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+func Metrics(c *gin.Context) {
+	
+	var payload metricsPayload
+
+	if err := c.ShouldBind(&payload); err != nil {
+		c.JSON(400, gin.H{
+			"code":    400,
+			"message": "Missing required parameters " + err.Error(),
+		})
+	}
```
Comment (russjones) [inline iot-meter/webhandlers/metrics_handler.go:30]: Should this return here? This will just pass through and execute the code below, same below.

**Review by russjones — COMMENTED** [review]
You probably didn't intend to check in `iot.db `.

File: iot-meter/iot_meter.go [hunk 4]
```diff
@@ -161,12 +33,16 @@ func main() {
 	}
 
 	// Load accounts database
-	_, err = db.LoadDatabase()
+	DB, err := db.LoadDatabase()
+	sqlDB, err := DB.DB()
 	if err != nil {
 		log.Fatal("Unable to open or migrate database")
```
Comment (awly) [inline iot-meter/iot_meter.go:39]:
```suggestion
		log.Fatal("Unable to open or migrate database: ", err)
```

File: iot-meter/iot_meter.go [hunk 5]
```diff
@@ -161,12 +33,16 @@ func main() {
...
 	}
+	defer sqlDB.Close()
 
 	// Run http server
-	setupServer()
-
+	err = runServer(DB)
+	if err != nil {
+		log.Fatal(err.Error())
```
Comment (awly) [inline iot-meter/iot_meter.go:46]:
```suggestion
		log.Fatal(err)
```

File: iot-meter/iot_meter.go [hunk 6]
```diff
@@ -161,12 +33,16 @@ func main() {
 	}
 
 	// Load accounts database
-	_, err = db.LoadDatabase()
+	DB, err := db.LoadDatabase()
```
Comment (awly) [inline iot-meter/iot_meter.go:36]: nit: `DB` -> `db` or `database`

File: iot-meter/models/account.go [hunk 7]
```diff
@@ -5,11 +5,36 @@ import (
...
+	}
+
+	return true
+}
+
+func (a *Account) IncreaseUserCount() {
+	a.TotalUsers += 1
+}
+
+func (a *Account) UpgradePlan() (bool) {
```
Comment (awly) [inline iot-meter/models/account.go:29]: Here and everywhere else - you don't need parentheses when there's a single return value

```suggestion
func (a *Account) UpgradePlan() bool {
```

File: iot-meter/models/account.go [hunk 8]
```diff
@@ -5,11 +5,36 @@ import (
...
+
+	if a.PlanType != "Basic" {
```
Comment (awly) [inline iot-meter/models/account.go:31]: nit: define constants for these plan types

File: iot-meter/webhandlers/login_handler.go [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+	
+		email := loginVals.Email
+		password := loginVals.Password
+		
+		// TEMPORARY CODE TO GENERATE JWT
+		if (email == "admin" && password == "admin") {
+			return &models.Account{
+				ID  : "MASTER",
+				Email : "admin@goteleport.com",
+				PlanType : "S",
```
Comment (awly) [inline iot-meter/webhandlers/login_handler.go:31]: your plan type should be either `Basic` or `Enterprise`.
see the comment above about defining constants for these

File: iot-meter/webhandlers/main_server.go [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+func (server *MainServer) Run() (error) {
+    err := server.listenTLS()
+    if err != nil {
+        return err
+    }
+
+    return nil
+}
```
Comment (awly) [inline iot-meter/webhandlers/main_server.go:176]: nit: rename `listenTLS` to `Run` and remove this wrapper method

File: iot-meter/webhandlers/metrics_handler.go [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+			"code":    500,
+			"message": "Fatal error creating new user " + tx.Error.Error(),
+		})
+	}
+
+	// Ir we got a not found, we have to insert, first check wether we can
+	if errors.Is(tx.Error, gorm.ErrRecordNotFound) {
+		
+
+		// Hopefully this will sync access to the db :)
```
Comment (awly) [inline iot-meter/webhandlers/metrics_handler.go:58]: the entire handler should probably run in a transaction, including the above two queries

File: iot-meter/webhandlers/metrics_handler.go [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+				"message": "Fatal error updating databass " + err.Error(),
+			})
+		}
+		
+	} else { // Otherwise just update TS, no need to make this locking
+
+		user.UpdatedAt = payload.Timestamp
+		tx := db.Save(user)
+
+		if tx.Error != nil && !errors.Is(tx.Error, gorm.ErrRecordNotFound) {
```
Comment (awly) [inline iot-meter/webhandlers/metrics_handler.go:105]: `ErrRecordNotFound` should also result in an error here

File: iot-meter/models/account.go [hunk 13]
```diff
@@ -5,11 +5,36 @@ import (
...
-	PlanType   string
-	TotalUsers uint
+	PlanType   string 	`gorm:"default:Basic" gorm:"not null"`
+	TotalUsers uint 	`gorm:"default:0" gorm:"not null"`
 	CreatedAt  time.Time
 	UpdatedAt  time.Time
 }
+
+func (a *Account) CanAcceptMoreUsers() (bool) {
+	if a.PlanType == "Basic" && a.TotalUsers >= 100 {
```
Comment (awly) [inline iot-meter/models/account.go:18]: `Enterprise` plan should be limited to 1000 users

File: iot-meter/webhandlers/metrics_handler.go [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+	tx = db.Where(&models.User{ID: payload.UserID, IotAccount: account}).First(&user)
+	if tx.Error != nil && !errors.Is(tx.Error, gorm.ErrRecordNotFound) {
+		c.JSON(500, gin.H{
+			"code":    500,
+			"message": "Fatal error creating new user " + tx.Error.Error(),
+		})
+	}
+
+	// Ir we got a not found, we have to insert, first check wether we can
+	if errors.Is(tx.Error, gorm.ErrRecordNotFound) {
```
Comment (alex-kovoy) [inline iot-meter/webhandlers/metrics_handler.go:55]: I am not sure what's happening here and what gets sent to the client. Are you forgetting the return statements?

**Pull request merged**